src/dune_rules/merlin/merlin.ml

```diff
@@ -1,9 +1,13 @@
 open Import
 
-let remove_extension file =
+let strip_pp_extensions file =
   let dir = Path.Build.parent_exn file in
-  let basename, _ext = String.lsplit2_exn (Path.Build.basename file) ~on:'.' in
-  Path.Build.relative dir basename
+  let basename = Path.Build.basename file in
+  match String.split basename ~on:'.' with
+  | [] | [ _ ] | [ _; _ ] -> file
+  | name :: exts ->
+    let ext = Option.value_exn (List.last exts) in
+    Path.Build.relative dir (sprintf "%s.%s" name ext)
```

Comment:
I still don't understand why you need to change this function and it's usage at line L253/R273.
Could you document its behavior and why it's needed ?

Reply:
This is to accommodate [this](https://github.com/andreypopp/dune/blob/10faea361ac82756ebd63a52d8ef66e45270923e/test/blackbox-tests/test-cases/merlin/server.t/run.t#L17-L20) test case.

Before this change we were tracking only config per module name (so extensions were stripped) but now, as we introduce readers which are per extension, we now track config per source file (with extensions).

We still want to treat `module_name.xx.xxx.ext` as `module_name.ext` though as the linked test case specifies, thus the function `strip_pp_extension` normalises a filename by stripping extensions added by a preprocessor.

Reply:
I think I finally understand. I initially though you had the same constraints as preprocessing specification which are per-module (that is, the same for the `ml` and the `mli` file of a module).

But readers can be different for the `ml` file and the `mli` file of a same module ?

Reply:
> But readers can be different for the ml file and the mli file of a same module ?

You mean for implementation and interface? Yes, I've raise the same question [here](https://github.com/ocaml/dune/pull/8567#issuecomment-1702857875).

But note though that for a single module an implementation might belong to a dialect while the interface could stay pure `.ml` — e.g. `.mlx` and `.mli`. So we still need to track config per file with extension even if we change `(merlin_reader)` to be same both for implementations and interfaces within a dialect.

Reply:
I am not very familiar with Merlin's handling of alternative readers, ~but it looks like the suffix map would allow to use a different reader for the interface and the implementation while keeping a unique configuration per module.~ In fact it's only available as a flag, not a configuration option. So having a configuration for each source file sounds like the easiest way to do it. 👍